src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+      }
+      std::cout << std::endl;
+      std::cout << "result=" << result << std::endl;
+    }
+    return result;
+  }
+
+  void PeakIntegrator::emg_vector(
+    const std::vector<double>& xs,
+    const double h,
+    const double mu,
+    const double sigma,
+    const double tau,
+    std::vector<double>& out_xs,
+    std::vector<double>& out_ys,
+    const bool compute_additional_points
+  ) const
+  {
+    out_xs = xs; // Copy all positions to output
+    for (const double x : out_xs) // For each x, estimate y
+    {
+      out_ys.push_back(emg_point(x, h, mu, sigma, tau));
+    }
+
+    if (!compute_additional_points) return;
+
+    // Compute the sampling step for the additional points
+    double avg_sampling { 0.0 };
+    for (Size i = 1; i < xs.size(); ++i)
+    {
+      avg_sampling += xs[i] - xs[i - 1];
+    }
+    avg_sampling /= xs.size() - 1;
+
+    // Stop adding points if the estimated y <= `est_y_threshold`
+    const double est_y_threshold { 1e-3 };
+
+    // Decide on which side the eventual additional points should be added
+    // The loop stops once the last added point's intensity is:
+    // - lower than the intensity on the opposite boundary
+    // - lower than `est_y_threshold`
+    if (out_ys.front() > out_ys.back())
+    {
+      const double target_intensity = out_ys.back();
+      while (out_ys.front() > target_intensity && out_ys.front() > est_y_threshold)
+      {
+        const double position = out_xs.front() - avg_sampling;
+        out_xs.insert(out_xs.begin(), position);
+        out_ys.insert(out_ys.begin(), emg_point(position, h, mu, sigma, tau));
+      }
+    }
+    else
+    {
+      const double target_intensity = out_ys.front();
+      while (out_ys.back() > target_intensity && out_ys.back() > est_y_threshold)
+      {
+        const double position = out_xs.back() + avg_sampling;
+        out_xs.push_back(position);
+        out_ys.push_back(emg_point(position, h, mu, sigma, tau));
+      }
+    }
+  }
+
+  double PeakIntegrator::computeInitialMean(
+    const std::vector<double>& xs,
+    const std::vector<double>& ys
+  ) const
+  {
+    const double max_intensity = *std::max_element(ys.begin(), ys.end());
+    // The intensity levels at which the mean candidates are computed
+    std::vector<double> percentages = { 0.6, 0.65, 0.7, 0.75, 0.8, 0.85 };
+    Size i = 0;
+    Size j = xs.size() - 1;
+    // Make sure left and right positions have an initial value
+    // This is to avoid situations (eg. cutoff peaks) where `max_intensity_threshold`
+    // is higher than the first point on a boundary of the peak. In such a case,
+    // the following nested loops would not get a chance to execute and the
+    // algorithm would fail
+    double left_pos = xs.front();
+    double right_pos = xs.back();
+    std::vector<double> mean_candidates;
+    for (const double height_percentage : percentages)
+    {
+      const double max_intensity_threshold = max_intensity * height_percentage;
```

Comment:
Good choice to use a relative height percentage.  If you can also note the following in the algorithm comment above:

"The avoidance of using the highest points of the peak apex also provides robustness to spurious points or random fluctuations in detector sampling from inflating the maximum peak height"